pkg/yang/types_builtin.go

```diff
@@ -506,32 +545,98 @@ func ParseNumber(s string) (n Number, err error) {
 		return n, errors.New("sign with no value")
 	}
 
+	n.Kind = Positive
+	ns := s
 	switch s[0] {
 	case '+':
-		s = s[1:]
+		ns = s[1:]
 	case '-':
 		n.Kind = Negative
-		s = s[1:]
+		ns = s[1:]
+	}
+	n.Value, err = strconv.ParseUint(ns, 0, 64)
+	if err == nil {
+		return n, nil
 	}
-	n.Value, err = strconv.ParseUint(s, 0, 64)
+
+	n.DecimalValue, n.FractionDigits, err = DecimalValueFromString(s, -1)
 	return n, err
 }
 
+// DecimalValueFromString returns an int64 representation of numStr and uint8
+// representation of the number of its fractional digits.
+// If fracDigRequired is >= 0, the number is represented with fracDigRequired
+// fractional digits, regardless of the precision of numStr, otherwise the
+// precision of numStr is used to set the number of fractional digits.
+// numStr must conform to Section 9.3.
+func DecimalValueFromString(numStr string, fracDigRequired int) (n *int64, fracDig *uint8, err error) {
+	s := numStr
+	decCount := strings.Count(s, ".")
```

Comment:
x := strings.Index(s, ".")
if x < 0 {not a number}

fd = len(s) - 1 - x
s = s[:x] + s[x+1:]
u, err := strconv.ParseUint(s)

if fd < fracDigRequired {
    m := 1
    for fd < fracDigRequired {
        m *= 10
        fd++
    }
    u *= m (probably should check for overflow)
}
if fd < fracDigRequired {
    d := 1
    for fd > fracDigRequired {
        d /= 10
        fd--
    }
    u /= d (probably should do rounding)
}